sdks/java/core/src/main/java/org/apache/beam/sdk/runners/inprocess/InMemoryWatermarkManager.java

```diff
@@ -366,22 +368,24 @@ public synchronized String toString() {
    * The input {@link TimeDomain#SYNCHRONIZED_PROCESSING_TIME} hold for an
    * {@link AppliedPTransform}.
    *
-   * <p>At any point, the hold value of an {@link SynchronizedProcessingTimeInputWatermark} is equal
-   * to the minimum across all pending bundles at the {@link AppliedPTransform} and all upstream
-   * {@link TimeDomain#SYNCHRONIZED_PROCESSING_TIME} watermarks. The value of the input
-   * synchronized processing time at any step is equal to the maximum of:
+   * <p>
+   * At any point, the hold value of an {@link SynchronizedProcessingTimeInputWatermark} is equal to
+   * the minimum across all pending elements at the {@link AppliedPTransform} and all upstream
+   * {@link TimeDomain#SYNCHRONIZED_PROCESSING_TIME} watermarks. The value of the input synchronized
+   * processing time at any step is equal to the maximum of:
    * <ul>
-   *   <li>The most recently returned synchronized processing input time
-   *   <li>The minimum of
-   *     <ul>
-   *       <li>The current processing time
-   *       <li>The current synchronized processing time input hold
-   *     </ul>
+   * <li>The most recently returned synchronized processing input time
+   * <li>The minimum of
+   * <ul>
+   * <li>The current processing time
+   * <li>the synchronized processing output timestamp of all currently pending elements</li>
+   * <li>all input {@link PValue} synchronized processing time watermarks</li>
+   * </ul>
    * </ul>
    */
   private static class SynchronizedProcessingTimeInputWatermark implements Watermark {
     private final Collection<? extends Watermark> inputWms;
-    private final Collection<CommittedBundle<?>> pendingBundles;
+    private final ListMultimap<WindowedValue<?>, Instant> pendingValueHolds;
```

Comment:
I assume the key here is a `WindowedValue<K>` representing the key and window with the hold? Add Javadoc on what the keys of this map are?